zebra-network/src/peer_set/candidate_set/tests/vectors.rs

```diff
@@ -115,6 +127,60 @@ fn rejects_all_addresses_if_applying_offset_causes_an_underflow() {
     assert!(validated_peers.next().is_none());
 }
 
+/// Test that calls to [`CandidateSet::update`] are rate limited.
+///
+/// This test is ignored by default because it takes about 31 seconds to run due to the 10 second
+/// rate limit interval.
+#[test]
+#[ignore]
+fn candidate_set_updates_are_rate_limited() {
+    let runtime = Runtime::new().expect("Failed to create Tokio runtime");
+    let _guard = runtime.enter();
+
+    let mut peer_request_tracker: VecDeque<_> =
+        iter::repeat(Instant::now()).take(GET_ADDR_FANOUT).collect();
+
+    let rate_limit_interval =
+        CandidateSet::<BoxService<Request, Response, BoxError>>::MIN_PEER_GET_ADDR_INTERVAL;
+
+    let peer_service = tower::service_fn(move |request| {
+        match request {
+            Request::Peers => {
+                // Get time from queue that the request is authorized to be sent
+                let authorized_request_time = peer_request_tracker
+                    .pop_front()
+                    .expect("peer_request_tracker should always have GET_ADDR_FANOUT elements");
+                // Check that the request was rate limited
+                assert!(Instant::now() >= authorized_request_time);
+                // Push a new authorization, updated by the rate limit interval
+                peer_request_tracker.push_back(authorized_request_time + rate_limit_interval);
+
+                // Return an empty list of peer addresses
+                future::ok(Response::Peers(vec![]))
+            }
+            _ => unreachable!("Received an unexpected internal message: {:?}", request),
+        }
+    });
+
+    let address_book = AddressBook::new(&Config::default(), Span::none());
+    let mut candidate_set = CandidateSet::new(Arc::new(Mutex::new(address_book)), peer_service);
+
+    runtime.block_on(async move {
+        let time_limit = Instant::now() + StdDuration::from_secs(31);
+
+        while Instant::now() <= time_limit {
+            candidate_set
+                .update()
+                .await
+                .expect("Call to CandidateSet::update should not fail");
+            // Avoid behaving too much like a spin-lock
+            sleep(StdDuration::from_millis(1)).await;
```

Comment:
Please wait for 1 second here, not one millisecond. We only expect to see events every few seconds, and spinning this fast will take CPU time away from other tests.

Reply:
Since I updated the test to use `tokio::time::advance`, I chose a better interval and added a comment describing why.